kauri/kauri.go

```diff
@@ -0,0 +1,301 @@
+			// cannot merge a and b if they both contain a contribution from the same ID.
+			if i == j {
+				canMerge = false
+			}
+			return canMerge
+		})
+		return canMerge
+	})
+	return canMerge
+}
+
+func (k *Kauri) verifyContribution(signature hotstuff.QuorumSignature, hash hotstuff.Hash) bool {
+	verified := false
+	block, ok := k.blockChain.Get(hash)
+	if !ok {
+		k.logger.Info("failed to fetch the block ", hash)
+		return verified
+	}
+	verified = k.crypto.Verify(signature, block.ToBytes())
+	return verified
+}
+
+func (k *Kauri) mergeWithContribution(currentSignature hotstuff.QuorumSignature) (bool, error) {
+	isVerified := k.verifyContribution(currentSignature, k.blockHash)
+	if !isVerified {
+		k.logger.Info("Contribution verification failed for view ", k.currentView,
+			"from participants", currentSignature.Participants(), " block hash ", k.blockHash)
+		return false, errors.New("unable to verify the contribution")
+	}
+	if k.aggregatedContribution == nil {
+		k.aggregatedContribution = currentSignature
+		return false, nil
+	}
+
+	if k.canMergeContributions(currentSignature, k.aggregatedContribution) {
+		new, err := k.crypto.Combine(currentSignature, k.aggregatedContribution)
+		if err == nil {
+			k.aggregatedContribution = new
+			if new.Participants().Len() >= k.configuration.QuorumSize() {
+				k.logger.Debug("Aggregated Complete QC and sending the event")
+				k.eventLoop.AddEvent(hotstuff.NewViewMsg{
+					SyncInfo: hotstuff.NewSyncInfo().WithQC(hotstuff.NewQuorumCert(
+						k.aggregatedContribution,
+						k.currentView,
+						k.blockHash,
+					)),
+				})
+				return true, nil
+			}
+		} else {
+			k.logger.Info("Failed to combine signatures: %v", err)
+			return false, errors.New("unable to combine signature")
+		}
+	} else {
+		k.logger.Debug("Failed to merge signatures due to overlap of signatures.")
+		return false, errors.New("unable to merge signature")
+	}
+	return false, nil
+}
+
+func isSubSet(a, b []hotstuff.ID) bool {
+	c := hotstuff.NewIDSet()
+	for _, id := range b {
+		c.Add(id)
+	}
+	for _, id := range a {
+		if !c.Contains(id) {
+			return false
+		}
+	}
+	return true
+}
+
+func (k *Kauri) randomizeIDS(hash hotstuff.Hash, leaderID hotstuff.ID) map[hotstuff.ID]int {
+	//assign leader to the root of the tree.
+
+	totalNodes := k.configuration.Len()
+	ids := make([]hotstuff.ID, 0, totalNodes)
+	for id := range k.configuration.Replicas() {
+		ids = append(ids, id)
+	}
+	sort.Slice(ids, func(i, j int) bool { return ids[i] < ids[j] })
+	seed := k.opts.SharedRandomSeed() + int64(binary.LittleEndian.Uint64(hash[:]))
+	//Shuffle the list of IDs using the shared random seed + the first 8 bytes of the hash.
+	rnd := rand.New(rand.NewSource(seed))
+	rnd.Shuffle(len(ids), reflect.Swapper(ids))
+	lIndex := 0
+	for index, id := range ids {
+		if id == leaderID {
+			lIndex = index
+		}
+	}
+	currentRoot := ids[0]
+	ids[0] = ids[lIndex]
+	ids[lIndex] = currentRoot
+	posMapping := make(map[hotstuff.ID]int)
```

Comment:
Why not just return the `ids` slice? Why do you want it as a map?

Since the hotstuff.IDs are 1-indexed, it would be possible to use a slice where the first entry is ignored, then you could do:
```go
idx := posMapping[myID]
```
To get the index of the `hotstuff.ID myID` in the same way as with a map.
Code would be like this:
```go
pos := make([]int, len(ids)+1)
for index, id := range ids {
	pos[id] = index
}
```